### PR TITLE
Feature: rail station class name filtering

### PR DESCRIPTION
**Pull request by perezdidac**
## Motivation / Problem

Players with many NewGRF enabled for train stations sometimes struggle at finding the right one to place in every situation. This PR introduces a filter edit box for quicker access when train station NewGRF are enabled in the game. This is especially helpful for those that care about station design and eye candy looking terminals.

## Description

Add a text box to the Build Rail Station window for filtering train station classes. This is similar to the previous feature implemented in the Object Selection window here: https://github.com/OpenTTD/OpenTTD/pull/8603.

This is what it looks like:

image

Note that the sorting logic is based on the added NewGRF order given the concerns in https://github.com/OpenTTD/OpenTTD/issues/8817:

image

## Limitations

No limitations.

## Testing

The following test cases have been tested:

- Re-opening the build train station window within the same game keeps the previously selected station class and station type.
- Starting/Loading a new game resets the selected station to the default one.
- Starting/Loading a new game with different NewGRFs.
- Default station always at the top.

All seems to work perfectly.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/rail_gui.cpp [hunk 1]
```diff
@@ -57,7 +61,7 @@ struct RailStationGUISettings {
 	Axis orientation;                 ///< Currently selected rail station orientation
 
 	bool newstations;                 ///< Are custom station definitions available?
-	StationClassID station_class;     ///< Currently selected custom station class (if newstations is \c true )
+	const StationClass *station_class; ///< Currently selected custom station class (if newstations is \c true )
```
Comment (stormcone) [inline src/rail_gui.cpp]: The comments here are aligned, maybe shift the other comments to right by one?

File: src/rail_gui.cpp [hunk 2]
```diff
@@ -891,6 +902,29 @@ struct BuildRailStationWindow : public PickerWindowBase {
...
 	Scrollbar *vscroll2;  ///< Vertical scrollbar of the matrix with new stations.
 
+	typedef GUIList<const StationClass *, StringFilter &> GUIStationClassList; ///< Type definition for the list to hold available station classes.
+
+	static const uint EDITBOX_MAX_SIZE = 16; ///< The maximum number of characters for the filter edit box.
+
+	static Listing   last_sorting;           ///< Default sorting of #GUIStationClassList.
+	static Filtering last_filtering;         ///< Default filtering of #GUIStationClassList.
+	static GUIStationClassList::SortFunction* const sorter_funcs[];   ///< Sort functions of the #GUIStationClassList.
+	static GUIStationClassList::FilterFunction* const filter_funcs[]; ///< Filter functions of the #GUIStationClassList.
```
Comment (stormcone) [inline src/rail_gui.cpp]:
```suggestion
	static GUIStationClassList::SortFunction * const sorter_funcs[];   ///< Sort functions of the #GUIStationClassList.
	static GUIStationClassList::FilterFunction * const filter_funcs[]; ///< Filter functions of the #GUIStationClassList.
```

File: src/rail_gui.cpp [hunk 3]
```diff
@@ -956,41 +1004,146 @@ struct BuildRailStationWindow : public PickerWindowBase {
...
+		this->InvalidateData();
 	}
 
 	virtual ~BuildRailStationWindow()
 	{
 		DeleteWindowById(WC_SELECT_STATION, 0);
 	}
 
+	/** Sort station classes by class name. */
+	static bool NameSorter(const StationClass *const &a, const StationClass *const &b)
```
Comment (stormcone) [inline src/rail_gui.cpp]:
```suggestion
	static bool NameSorter(const StationClass * const &a, const StationClass * const &b)
```

File: src/rail_gui.cpp [hunk 4]
```diff
@@ -956,41 +1004,146 @@ struct BuildRailStationWindow : public PickerWindowBase {
...
+		GetString(buffer_a, a->name, lastof(buffer_a));
+
+		char buffer_b[DRAW_STRING_BUFFER];
+		GetString(buffer_b, b->name, lastof(buffer_b));
+
+		return strnatcmp(buffer_a, buffer_b, true) < 0; // Sort by name (natural sorting).
+	}
+
+	/** Filter station classes by class name. */
+	static bool CDECL TagNameFilter(const StationClass *const *oc, StringFilter& filter)
```
Comment (stormcone) [inline src/rail_gui.cpp]:
```suggestion
	static bool CDECL TagNameFilter(const StationClass * const *oc, StringFilter &filter)
```

File: src/rail_gui.cpp [hunk 5]
```diff
@@ -1367,6 +1509,17 @@ struct BuildRailStationWindow : public PickerWindowBase {
 	}
 };
 
+Listing BuildRailStationWindow::last_sorting = { false, 0 };
+Filtering BuildRailStationWindow::last_filtering = { false, 0 };
+
+BuildRailStationWindow::GUIStationClassList::SortFunction* const BuildRailStationWindow::sorter_funcs[] = {
```
Comment (stormcone) [inline src/rail_gui.cpp]:
```suggestion
BuildRailStationWindow::GUIStationClassList::SortFunction * const BuildRailStationWindow::sorter_funcs[] = {
```

File: src/rail_gui.cpp [hunk 6]
```diff
@@ -1367,6 +1509,17 @@ struct BuildRailStationWindow : public PickerWindowBase {
...
+	&NameSorter,
+};
+
+BuildRailStationWindow::GUIStationClassList::FilterFunction* const BuildRailStationWindow::filter_funcs[] = {
```
Comment (stormcone) [inline src/rail_gui.cpp]:
```suggestion
BuildRailStationWindow::GUIStationClassList::FilterFunction * const BuildRailStationWindow::filter_funcs[] = {
```

File: src/rail_gui.cpp [hunk 7]
```diff
@@ -169,13 +173,20 @@ static void PlaceRail_Waypoint(TileIndex tile)
 	}
 }
 
+/**
+ * Helper macro to return the StationClassID of a given StationClass instance. This is needed in order
+ * to provide the StationClassID to specific methods that require it like to draw the sprites and send
+ * the build station command over the network.
+ */
+#define GET_STATION_CLASS_ID(X) StationClass::Allocate(X->global_id)
```
Comment (LordAro) [inline src/rail_gui.cpp]: This feels like the wrong way round. Why not continue to use StationClassID and use StationClass::Get where needed? Seems like fewer changes would be needed overall

**Review by stormcone — COMMENTED** [review]
Some minor code style fixes. Maybe @SamuXarick could take a look at it also. :)

@Devs: Please add the preview label to this. :) (And thanks for that feature. :P)

File: src/rail_gui.cpp (hunk 7 above)
Comment (perezdidac) [reply src/rail_gui.cpp]: Hi LordAro and thanks for your quick review!
I am flipping things around and yep you are right! For some reason I though I needed to hold a pointer to the object in GUIStationClassList. I am almost done pushing an update, will resolve the comment once all works fine.

**Comment by stormcone** [comment]
Would it be possible that the filter textbox get automatically focused when the station building window opens? So we could type in immediately without to click on the textbox. Just like in the case of the build object window.

**Comment by perezdidac** [comment]
> Would it be possible that the filter textbox get automatically focused when the station building window opens? So we could type in immediately without to click on the textbox. Just like in the case of the build object window.

Done! All good.

File: src/rail_gui.cpp [hunk 8]
```diff
@@ -891,6 +895,30 @@ struct BuildRailStationWindow : public PickerWindowBase {
...
+	QueryString filter_editbox;              ///< Filter editbox.
+
+	/** Scroll #WID_BRAS_NEWST_SCROLL so that the selected station class is visible. */
+	void EnsureSelectedStationClassIsVisible()
+	{
+		uint pos = 0;
+		for (auto station_class : this->station_classes) {
+			if (station_class == _railstation.station_class) break;
+			pos++;
+		}
```
Comment (LordAro) [inline src/rail_gui.cpp:923]:
```suggestion
		uint pos = std::find(this->station_classes.begin(), this->station_classes.end(), _railstation.station_class) - this->station_classes.begin();
```
Though I'm undecided whether this is actually better. Might need some "was class actually found" checking

File: src/rail_gui.cpp [hunk 9]
```diff
@@ -956,37 +998,144 @@ struct BuildRailStationWindow : public PickerWindowBase {
...
+	void SelectClassAndStation()
+	{
+		if (_railstation.station_class == StationClassID::STAT_CLASS_DFLT) {
+			/* This happens during the first time the window is open during the game life cycle. */
+			this->SelectOtherClass(StationClassID::STAT_CLASS_DFLT);
+		} else {
+			/* Check if the previously selected station class is not available anymore as a
+			 * result of starting a new game without the corresponding NewGRF. */
+			bool available = false;
+			for (uint i = 0; StationClass::GetClassCount(); ++i) {
```
Comment (LordAro) [inline src/rail_gui.cpp]: this is missing a condition

**Review by LordAro — DISMISSED** [review]

**Comment by perezdidac** [comment]
Please do not merge yet, since https://github.com/OpenTTD/OpenTTD/issues/8817 was filed I am considering not sorting alphabetical here either.

**Comment by stormcone** [comment]
I would be happy with a drop down menu with the sort orders here too. :)

**Comment by perezdidac** [comment]
> I would be happy with a drop down menu with the sort orders here too. :)

Yep that could be a good addition, however I believe it might be better to do it in a separate PR. I am also thinking about ´favoriting´ / ´starring´ classes or objects to being able to quickly use the most frequent ones,

File: src/rail_gui.cpp [hunk 10]
```diff
@@ -956,37 +1004,131 @@ struct BuildRailStationWindow : public PickerWindowBase {
...
+				if ((StationClassID)i == _railstation.station_class) {
+					available = true;
+					break;
+				}
+			}
+
+			if (available) {
+				this->SelectOtherClass(_railstation.station_class);
+			} else {
+				this->SelectOtherClass(this->station_classes[0]);
```
Comment (LordAro) [inline src/rail_gui.cpp]: I assume there's no possibility of station_classes being empty? Maybe a comment/assert to "verify" it :)

File: src/rail_gui.cpp [hunk 11]
```diff
@@ -956,37 +1004,131 @@ struct BuildRailStationWindow : public PickerWindowBase {
...
+			}
+		}
+
+		assert(StationClass::Get(_railstation.station_class)->GetUISpecCount() > 0); // station GUI should be disabled elsewise
```
Comment (LordAro) [inline src/rail_gui.cpp]: would be more normal to put this at the top of the function, imo - fail as early as possible (assuming i'm reading it right and the above doesn't change `_railstation.station_class` at all)

**Review by LordAro — COMMENTED** [review]
More or less fine, afaict. Looks good!

File: src/rail_gui.cpp (hunk 11 above)
Comment (perezdidac) [reply src/rail_gui.cpp]: This got copied from my work on the object selection window, this doesn't apply here at all. removing, thank you!

File: src/rail_gui.cpp (hunk 10 above)
Comment (perezdidac) [reply src/rail_gui.cpp]: Changing to StationClassID::STAT_CLASS_DFLT if not available as it's always the 1st in the list. Thanks!

**Review by LordAro — APPROVED** [review]

**Comment by Kuhnovic** [comment]
I'm playing the latest release candidate with a bunch of friends as we speak. Myself and several others are having issues with this feature. The filter works fine, but auto focus on the text field is a problem. For example, if you want to remove a station tile, you typically click the station button and hit R. But that R button is now simply adding an r to the filter string. Same thing if you press delete to clear all windows... you're deleting characters in the filter string. This really kills the whole flow of the UI, it's quite annoying.

One of my friends pointed out that the Town Directory window has a similar filter string, but this text field does not automatically get the focus. And then it works just fine. So I think removing the auto focus should do the trick.

**Comment by LordAro** [comment]
You raise good points. I'd recommend raising a new issue though. Comments on closed/merged PRs tend to get lost

**Comment by Kuhnovic** [comment]
Good point, I'll do it right away!

**Comment by stormcone** [comment]
If you press the ESC, the focus is lost and you can use the hotkeys.

**Pull request merged**